src/include/daos_srv/bio.h

```diff
@@ -219,26 +219,33 @@ uint16_t bio_iov2media(const struct bio_iov *biov)
 static inline int
 bio_sgl_init(struct bio_sglist *sgl, unsigned int nr)
 {
-	memset(sgl, 0, sizeof(*sgl));
-
+	sgl->bs_nr_out = 0;
 	sgl->bs_nr = nr;
+
+	if (nr == 0) {
```

Comment:
do we want to return 0 here?   What did it do before?  Per malloc man page, malloc of 0 size may return NULL or may return something else that can be passed to free.  The question here is whether this should either assert or return -DER_INVAL or some such.  In the old code, if D_ALLOC_ARRAY returned NULL, it would have gotten -DER_NOMEM

Reply:
I would like to have the similar behavior in all places of this implementation. From my point of view, bio_sgl_init() is the same as d_sgl_init() but I don't mergem them yet. In original daos_sgl_init() I get this behavior and propagate it to all places (made them similar). I don't know what was the reason to have this in original implementation and I don't have strong reason to change it.